### PR TITLE
[consensus] Update Sui Interface for Mysticeti

### PR DESCRIPTION
**Pull request by arun-koshy**
- Plumb committee/params/protocol config through to AuthorityNode
- Add basic transaction validation
- Plug in mysticeti transaction submission client for Sui
- Plug in consensus output channel to Sui

### CODE REVIEW

**Comment by akichidis** [comment]
> Sharing where I am currently at with Sui Interface
> 
> Need to rethink approach of starting AuthorityNode without having to expose NetworkManager trait outside of consensus crate

I haven't looked over the whole PR, but that was a same-ish issue I had to deal with the Mysticeti prototype codebase. I think the are a few options to get around this:
* introduce separate methods to start the AuthorityNode production/test where the trait implementations are injected as part of the method and the corresponding NetworkManager implementation gets initialised. For example in the Mysticeti prototype I ended up introducing  two separate methods for start in production/benchmarking https://github.com/MystenLabs/mysticeti/blob/3b78b12af94986abbced89e662eedf157248b465/mysticeti-core/src/validator.rs#L165 which would inject the corresponding trait implementations - so I won't have to propagate to mysticeti_manager.
* Have two different AuthorityNode implementations 
* Avoid the use of generics for now at the AuthorityNode (that could be the worst case solution)

**Comment by arun-koshy** [comment]
> introduce separate methods to start the AuthorityNode production/test where the trait implementations are injected as part of the method and the corresponding NetworkManager implementation gets initialised. For example in the Mysticeti prototype I ended up introducing two separate methods for start in production/benchmarking https://github.com/MystenLabs/mysticeti/blob/3b78b12af94986abbced89e662eedf157248b465/mysticeti-core/src/validator.rs#L165 which would inject the corresponding trait implementations - so I won't have to propagate to mysticeti_manager.


Awesome I think I am going to go this route! Will also have a brief chat with @mwtian about it in a little bit but this seems like a good path forward to me

**Comment by arun-koshy** [comment]
@mwtian had a great suggestion of just defining a new type that would allow me not to have to propagate the trait information through to Sui.

Cleaned up most of the PR so it should be ready for review, left some TODOs that you guys can take a look at and see some of the open questions/items left to do. Let me know if any are not valid or can be removed/fixed in this PR. 

Thanks!

File: consensus/core/src/context.rs [hunk 1]
```diff
@@ -20,7 +20,7 @@ use consensus_config::{NetworkKeyPair, ProtocolKeyPair};
 /// of this authority.
 #[allow(dead_code)]
 #[derive(Clone)]
-pub(crate) struct Context {
+pub struct Context {
```
Comment (mwtian) [inline consensus/core/src/context.rs:23]: Are we switching all `pub(crate)` to `pub`? It is fine but we just need to be consistent.

File: consensus/core/src/lib.rs [hunk 2]
```diff
@@ -26,3 +26,8 @@ mod universal_committer;
 mod leader_timeout;
 #[cfg(test)]
 mod test_dag;
+
+pub use authority_node::ConsensusAuthority;
```
Comment (mwtian) [inline consensus/core/src/lib.rs:30]: This list of public types make sense.

File: consensus/core/src/network/mod.rs [hunk 3]
```diff
@@ -61,6 +61,8 @@ where
     fn install_service(&self, network_keypair: NetworkKeyPair, service: Arc<S>);
 
     /// Stops the network service.
+    // TODO: Investigate if we need to use async fn here
+    #[allow(async_fn_in_trait)]
```
Comment (mwtian) [inline consensus/core/src/network/mod.rs:65]: Why is this attribute needed?

File: consensus/core/src/core.rs [hunk 4]
```diff
@@ -409,10 +416,20 @@ mod test {
         let context = Arc::new(context);
         let store = Arc::new(MemStore::new());
         let dag_state = Arc::new(RwLock::new(DagState::new(context.clone(), store.clone())));
-        let block_manager = BlockManager::new(context.clone(), dag_state);
+        let block_manager = BlockManager::new(context.clone(), dag_state.clone());
         let (_transaction_client, tx_receiver) = TransactionClient::new(context.clone());
         let transaction_consumer = TransactionConsumer::new(tx_receiver, context.clone(), None);
 
+        #[allow(clippy::disallowed_methods)] // allow unbounded_channel()
```
Comment (mwtian) [inline consensus/core/src/core.rs:423]: Can we allow list this clippy warning just for `consensus/`?

File: crates/sui-core/src/consensus_manager/mysticeti_manager.rs [hunk 5]
```diff
@@ -107,81 +102,69 @@ impl ConsensusManagerTrait for MysticetiManager {
             return;
         };
 
-        let parameters = mysticeti_parameters(&committee);
-        let committee = mysticeti_committee(&committee);
+        let parameters = Parameters {
```
Comment (mwtian) [inline crates/sui-core/src/consensus_manager/mysticeti_manager.rs:105]: Add a TODO to fill in other fields.

File: crates/sui-core/src/consensus_types/consensus_output_api.rs [hunk 6]
```diff
@@ -112,18 +113,18 @@ impl ConsensusOutputAPI for narwhal_types::ConsensusOutput {
...
         // TODO: Implement this in Mysticeti.
         None
     }
 
     fn leader_round(&self) -> u64 {
-        self.anchor.round
+        self.leader.round as u64
     }
 
     fn leader_author_index(&self) -> AuthorityIndex {
```
Comment (mwtian) [inline crates/sui-core/src/consensus_types/consensus_output_api.rs:126]: Let's change AuthorityIndex to u32.

File: crates/sui-core/src/consensus_types/consensus_output_api.rs [hunk 7]
```diff
@@ -132,18 +133,19 @@ impl ConsensusOutputAPI for mysticeti_core::consensus::linearizer::CommittedSubD
...
+        self.commit_index
     }
 
     fn transactions(&self) -> ConsensusOutputTransactions {
         self.blocks
             .iter()
             .map(|block| {
                 let round = block.round();
-                let author = block.author() as AuthorityIndex;
+                let author = block.author().value() as u16;
```
Comment (mwtian) [inline crates/sui-core/src/consensus_types/consensus_output_api.rs:144]: Why not `as AuthorityIndex`?

File: crates/sui-core/src/consensus_validator.rs [hunk 8]
```diff
@@ -147,6 +148,27 @@ impl TransactionValidator for SuiTxValidator {
     }
 }
 
+impl TransactionVerifier for SuiTxValidator {
+    fn verify_batch(
+        &self,
+        _protocol_config: &ProtocolConfig,
+        batch: &[&[u8]],
+    ) -> Result<(), ValidationError> {
+        // TODO(mysticeti): ensure we are properly verifying transactions
```
Comment (mwtian) [inline crates/sui-core/src/consensus_validator.rs:157]: Mention what are done improperly here?

File: crates/sui-core/src/mysticeti_adapter.rs [hunk 9]
```diff
@@ -101,18 +74,22 @@ impl SubmitToConsensus for LazyMysticetiClient {
         transaction: &ConsensusTransaction,
         _epoch_store: &Arc<AuthorityPerEpochStore>,
     ) -> SuiResult {
-        // The retrieved MysticetiClient can be from the past epoch. Submit would fail after
+        // TODO(mysticeti): confirm comment is still true
+        // The retrieved TransactionClient can be from the past epoch. Submit would fail after
         // Mysticeti shuts down, so there should be no correctness issue.
         let client = self.get().await;
+        let tx_bytes = bcs::to_bytes(&transaction).expect("Serialization should not fail.");
+        // TODO(mysticeti): convert consensus::ClientError -> SuiError properly
```
Comment (mwtian) [inline crates/sui-core/src/mysticeti_adapter.rs:82]: What is the issue with the current approach?

**Review by mwtian — APPROVED** [review]
LGTM structure-wise.

**Comment by akichidis** [comment]
> @mwtian had a great suggestion of just defining a new type that would allow me not to have to propagate the trait information through to Sui.
> 
> Cleaned up most of the PR so it should be ready for review, left some TODOs that you guys can take a look at and see some of the open questions/items left to do. Let me know if any are not valid or can be removed/fixed in this PR.
> 
> Thanks!

Oh yes even better this way! 👍

File: crates/sui-core/src/consensus_manager/mysticeti_manager.rs [hunk 10]
```diff
@@ -107,81 +102,69 @@ impl ConsensusManagerTrait for MysticetiManager {
...
-                        );
-                    }
-
-                    tracing::error!("Failed starting Mysticeti, retry {}: {:?}", retries, err);
-                    tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
-                }
-            }
-        }
+        // TODO: that should be replaced by a metered channel. We can discuss if unbounded approach
+        // is the one we want to go with.
```
Comment (akichidis) [inline crates/sui-core/src/consensus_manager/mysticeti_manager.rs:124]: 👍

File: crates/sui-core/src/consensus_manager/mysticeti_manager.rs [hunk 11]
```diff
@@ -107,81 +102,69 @@ impl ConsensusManagerTrait for MysticetiManager {
...
+
+        let consensus_handler = consensus_handler_initializer.new_consensus_handler();
+        let consumer = CommitConsumer::new(
+            commit_sender,
+            // TODO(mysticeti): remove dependency on narwhal executor
+            consensus_handler.last_executed_sub_dag_index().await,
+        );
+
+        // TODO(mysticeti): Investigate if we need to return potential errors from
+        // AuthorityNode and add retries here?
```
Comment (akichidis) [inline crates/sui-core/src/consensus_manager/mysticeti_manager.rs:136]: The main reason for adding retries is due to of transient errors having to do with some resource allocation from previous that didn't get released successfully yet. For instance we saw such examples during network shutdown where the port wasn't released yet while trying to spin up the new instance, but it rather needed a few millis before do so. I think having the retries is ok but we can discuss about it.

File: crates/sui-types/src/sui_system_state/epoch_start_sui_system_state.rs [hunk 12]
```diff
@@ -182,6 +184,23 @@ impl EpochStartSystemStateTrait for EpochStartSystemStateV1 {
         committee_builder.build()
     }
 
+    #[allow(clippy::mutable_key_type)]
+    fn get_mysticeti_committee(&self) -> ConsensusCommittee {
```
Comment (akichidis) [inline crates/sui-types/src/sui_system_state/epoch_start_sui_system_state.rs:188]: Nice 👍

File: consensus/core/src/network/mod.rs (hunk 3 above)
Comment (akichidis) [reply consensus/core/src/network/mod.rs:65]: hhm same question from my side as well

**Review by akichidis — APPROVED** [review]
Nice 💯  . Changes make sense. Similar comment from my side regarding switching from pub(crate) to pub. I think overall pub(crate) makes sense to use for structs that we don't want to leak and it's good practice to use as a default option , so basically to force our selfs switching to pub only when intentionally need to increase the visibility. Working almost as a sanity check.

File: consensus/core/src/network/mod.rs (hunk 3 above)
Comment (arun-koshy) [reply consensus/core/src/network/mod.rs:65]: I do as the linter tells me 🙇🏽 

This was the warning was a result of changing the visibility for `NetworkManager`. The todo was to see if we should go with the suggestion in the linter or another approach.

```
mod.rs(66, 5): original diagnostic
use of `async fn` in public traits is discouraged as auto trait bounds cannot be specified
you can suppress this lint if you plan to use the trait only in your own code, or do not care about auto traits like `Send` on the `Future`
`#[warn(async_fn_in_trait)]` on by default

mod.rs(66, 5): you can alternatively desugar to a normal `fn` that returns `impl Future` and add any desired bounds such as `Send`, but these cannot be relaxed without a breaking API change: ` -> impl std::future::Future<Output = ()> + Send`
```

File: crates/sui-core/src/consensus_validator.rs (hunk 8 above)
Comment (arun-koshy) [reply crates/sui-core/src/consensus_validator.rs]: Forgot to remove the todo, I previously just had this return Ok(). Should be good now as its doing the same thing as the nw tx validator.

File: crates/sui-core/src/mysticeti_adapter.rs (hunk 9 above)
Comment (arun-koshy) [reply crates/sui-core/src/mysticeti_adapter.rs]: Forgot to remove this TODO as well

File: crates/sui-core/src/consensus_manager/mysticeti_manager.rs (hunk 10 above)
Comment (mwtian) [reply crates/sui-core/src/consensus_manager/mysticeti_manager.rs:125]: I think unbounded is fine. But we should build a wrapper to meter / monitor unbounded channels too. Adding a work item.

**Pull request merged**